lib/ruled/client.lua

```diff
@@ -0,0 +1,717 @@
+            value = value(c, props)
+        end
+
+        local ignore = module.high_priority_properties[property] or
+            module.delayed_properties[property] or force_ignore[property]
+
+        if not ignore then
+            if module.extra_properties[property] then
+                module.extra_properties[property](c, value, props)
+            elseif type(c[property]) == "function" then
+                c[property](c, value)
+            else
+                c[property] = value
+            end
+        end
+    end
+
+    -- Apply all callbacks.
+    if callbacks then
+        for _, callback in pairs(callbacks) do
+            protected_call(callback, c)
+        end
+    end
+
+    -- Apply the delayed properties
+    for prop, handler in pairs(module.delayed_properties) do
+        if not force_ignore[prop] then
+            local value = props[prop]
+            if value ~= nil then
+                if type(value) == "function" then
+                    value = value(c, props)
+                end
+                handler(c, value, props)
+            end
+        end
+    end
+
+    -- Do this at last so we do not erase things done by the focus signal.
+    if props.focus and (type(props.focus) ~= "function" or props.focus(c)) then
+        c:emit_signal('request::activate', "rules", {raise=not capi.awesome.startup})
+    end
+end
+
+function module.execute(...) crules:_execute(...) end
+
+-- TODO v5 deprecate this
+function module.completed_with_payload_callback(c, props, callbacks)
+    module.execute(c, props, callbacks)
+end
+
+gobject._setup_class_signals(module)
+
+capi.client.connect_signal("request::manage", module.apply)
+
+-- Request rules to be added **after** all modules are loaded, but before the
+-- clients are managed. This allows module to listen to rules being added and
+-- either modify them or add their own in the right order.
+local function request_rules()
+    module.emit_signal("request::rules")
+end
+
+capi.client.connect_signal("scanning", request_rules)
+
+--@DOC_rule_COMMON@
+
+return setmetatable(module, {
+    __newindex = function(_, k, v)
+        if k == "rules" then
+            gdebug.deprecate(
+                "Use ruled.client.append_rules instead awful.rules.rules",
+                {deprecated_in=5}
+            )
+
+            -- Clearing the rule was supported, so it still has to be. This is
+            -- a bad idea. There is no plan to make this API public.
+            if not next(v) then
+                -- It isn't possible to just set it to {}, there is other
+                -- references to the table.
+                for k2 in pairs(crules._matching_rules["awful.rules"]) do
+                    crules._matching_rules["awful.rules"][k2] = nil
+                end
+            else
+                crules:append_rules("awful.rules", v)
+            end
+        else
+            rawset(k, v)
```

Comment:
@Elv13 
This `rawset` appears to be wrong.

I am seeing an error because of this, which probably at least could be more helpful / meaningful - if it cannot be fixed to "just keep working":

```
error while running function!
stack traceback:
        /usr/share/awesome/lib/ruled/client.lua:894: in function '__newindex'
        /home/user/.config/awesome/lib/stop_unfocused.lua:53: in main chunk
        [C]: in function 'require'
        /home/user/.config/awesome/rc.lua:170: in main chunk
error: /usr/share/awesome/lib/ruled/client.lua:894: bad argument #1 to 'rawset' (table expected, got string)
```

This is triggered through some monkeypatching / manually adding a function to `awful.rules`:

```lua
local awful = require('awful')
if not awful.rules.get_props_and_callbacks then
function awful.rules.get_props_and_callbacks(c, _rules)
    local matching_rules = awful.rules.matching_rules(c, _rules)
    if #matching_rules == 0 then
        return
    end

    local props = {}
    local callbacks = {}
    for _, entry in ipairs(matching_rules) do
        gears.table.crush(props, entry.properties or {})
        if entry.callback then
            table.insert(callbacks, entry.callback)
        end
    end
    return props, callbacks
end
end
```